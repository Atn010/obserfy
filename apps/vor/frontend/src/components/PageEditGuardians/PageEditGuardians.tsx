import React, { FC, useState } from "react"
import BackNavigation from "../BackNavigation/BackNavigation"
import {
  Guardians,
  useGetSchoolGuardians,
} from "../../api/useGetSchoolGuardians"
import { Student, useGetStudent } from "../../api/useGetStudent"
import { Typography } from "../Typography/Typography"
import { Card } from "../Card/Card"
import { NEW_GUARDIANS_URL, STUDENT_PROFILE_URL } from "../../routes"
import { Box } from "../Box/Box"
import SearchBar from "../SearchBar/SearchBar"
import Icon from "../Icon/Icon"
import { ReactComponent as PlusIcon } from "../../icons/plus.svg"
import { ReactComponent as LinkIcon } from "../../icons/link.svg"
import { ReactComponent as RemoveIcon } from "../../icons/close.svg"
import { Flex } from "../Flex/Flex"
import GuardianRelationshipPickerDialog from "../GuardianRelationshipPickerDialog/GuardianRelationshipPickerDialog"
import { usePostGuardianRelation } from "../../api/usePostGuardianRelation"
import { Link } from "../Link/Link"
import { Button } from "../Button/Button"
import { useDeleteGuardianRelation } from "../../api/useDeleteGuardianRelation"
import Dialog from "../Dialog/Dialog"
import DialogHeader from "../DialogHeader/DialogHeader"
import LoadingIndicator from "../LoadingIndicator/LoadingIndicator"

interface Props {
  studentId: string
}
export const PageEditGuardians: FC<Props> = ({ studentId }) => {
  const [searchTerm, setSearchTerm] = useState("")
  const student = useGetStudent(studentId)
  const guardians = useGetSchoolGuardians()

  const filteredGuardians = guardians.data?.filter((guardian) => {
    return (
      guardian.name.match(new RegExp(searchTerm, "i")) &&
      !student.data?.guardians.map(({ id }) => id).includes(guardian.id)
    )
  })

  return (
    <Box maxWidth="maxWidth.md" mx="auto">
      <BackNavigation
        to={STUDENT_PROFILE_URL(studentId)}
        text="Student Profile"
      />
      <Typography.H5 mx={3} mb={4} mt={3}>
        {student.data?.name}
      </Typography.H5>

      <Typography.Body mx={3} mb={2} color="textMediumEmphasis">
        Current guardians
      </Typography.Body>
      {(student.data?.guardians.length ?? 0) === 0 && (
        <Card borderRadius={[0, "default"]} mb={3} mx={[0, 3]}>
          <Typography.Body m={3} color="textMediumEmphasis" fontSize={1}>
            No guardians set yet
          </Typography.Body>
        </Card>
      )}
      {student.data?.guardians.map(
        (guardian) =>
          guardian && (
            <CurrentGuardiansCard guardian={guardian} studentId={studentId} />
          )
      )}
      <Card borderRadius={[0, "default"]} mb={2} mx={[0, 3]}>
        <Link to={NEW_GUARDIANS_URL(studentId)}>
          <Flex alignItems="center" p={3}>
            <Typography.Body lineHeight={1} fontSize={1}>
              Create new guardian
            </Typography.Body>
            <Icon
              as={LinkIcon}
              m={0}
              ml="auto"
              mr={2}
              fill="textMediumEmphasis"
            />
          </Flex>
        </Link>
      </Card>
      <Typography.Body mx={3} mt={4} color="textMediumEmphasis">
        Add existing guardian
      </Typography.Body>
      <Box px={3} pb={3} pt={2}>
        <SearchBar
          width="100%"
          value={searchTerm}
          onChange={(e) => setSearchTerm(e.target.value)}
        />
      </Box>
      {filteredGuardians?.map((guardian) => {
        return (
          <OtherGuardiansCard
            key={guardian.id}
            studentId={studentId}
            guardian={guardian}
          />
        )
      })}
    </Box>
  )
}

const CurrentGuardiansCard: FC<{
  guardian: Student["guardians"][0]
  studentId: string
}> = ({ guardian, studentId }) => {
  const [showDialog, setShowDialog] = useState(false)
  const [mutate, { error, status }] = useDeleteGuardianRelation(
    guardian.id,
    studentId
  )

  return (
    <>
      <Card borderRadius={[0, "default"]} mb={2} mx={[0, 3]}>
        <Flex alignItems="center">
          <Box p={3}>
            <Typography.Body lineHeight={1} mb={2}>
              {guardian.name}
            </Typography.Body>
            <Typography.Body
              lineHeight={1}
              color="textMediumEmphasis"
              fontSize={1}
            >
              {guardian.email || "No email"}
            </Typography.Body>
          </Box>
          <Button
            mr={3}
            ml="auto"
            variant="outline"
            px={2}
            onClick={() => setShowDialog(true)}
          >
            <Icon as={RemoveIcon} m={0} fill="danger" />
          </Button>
        </Flex>
      </Card>
      {showDialog && (
        <Dialog>
          <Flex
            alignItems="center"
            backgroundColor="surface"
            py={1}
            sx={{ position: "relative" }}
          >
            <Typography.H6
              width="100%"
              sx={{
                position: "absolute",
                pointerEvents: "none",
                textAlign: "center",
                alignContent: "center",
              }}
            >
              Remove Guardian?
            </Typography.H6>
            <Button
              variant="secondary"
              my={1}
              onClick={() => setShowDialog(false)}
              ml={2}
            >
              Cancel
            </Button>
            <Button
              variant="secondary"
              ml="auto"
              my={1}
              color="danger"
              onClick={async () => {
                const result = await mutate()
                if (!error && result?.status === 201) {
                  setShowDialog(false)
                }
              }}
              mr={2}
            >
              {status === "loading" && <LoadingIndicator />} Remove
            </Button>
          </Flex>
          <Typography.Body p={3}>
            Do you really want to remove {guardian.name}?
          </Typography.Body>
        </Dialog>
      )}
    </>
  )
}

const OtherGuardiansCard: FC<{ guardian: Guardians; studentId: string }> = ({
  guardian,
  studentId,
}) => {
  const [showDialog, setShowDialog] = useState(false)
  const [mutate, { error, status }] = usePostGuardianRelation(
    guardian,
    studentId
  )

  return (
    <>
      <Card
        borderRadius={[0, "default"]}
        mb={2}
        mx={[0, 3]}
        onClick={() => setShowDialog(true)}
        sx={{
          cursor: "pointer",
        }}
      >
        <Flex alignItems="center">
          <Icon as={PlusIcon} m={0} ml={3} fill="primary" />
          <Box p={3}>
            <Typography.Body lineHeight={1} mb={2}>
              {guardian.name}
            </Typography.Body>
            <Typography.Body
              lineHeight={1}
              color="textMediumEmphasis"
              fontSize={1}
            >
              {guardian.email || "No email"}
            </Typography.Body>
          </Box>
        </Flex>
      </Card>
      {showDialog && (
        <GuardianRelationshipPickerDialog
          loading={status === "loading"}
          onDismiss={() => setShowDialog(false)}
          onAccept={async (relationship) => {
            const result = await mutate(relationship)
            if (!error && result?.status === 201) {
              setShowDialog(false)
            }
          }}
        />
      )}
    </>
  )
}

export default PageEditGuardians
