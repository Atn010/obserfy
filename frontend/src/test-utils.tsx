/* eslint-disable import/no-extraneous-dependencies */
import React, { FC, ReactElement } from "react"
import { ThemeProvider } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import { IntlProvider } from "gatsby-plugin-intl3"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import Theme from "./gatsby-plugin-theme-ui"
import Box from "./components/Box/Box"
import en from "./intl/en.json"

const AllProviders: FC = ({ children }) => {
  // TODO: Maybe this better be outside
  const history = createHistory(createMemorySource("/"))
  return (
    <LocationProvider history={history}>
      <ThemeProvider theme={Theme}>
        <IntlProvider locale="en" messages={en}>
          <Box fontSize={[16, 20]}>{children}</Box>
        </IntlProvider>
      </ThemeProvider>
    </LocationProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
