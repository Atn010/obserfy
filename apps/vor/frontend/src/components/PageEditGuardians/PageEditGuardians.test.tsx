import React from "react"
import { render } from "../../test-utils"
import { Basic } from "./PageEditGuardians.stories"

describe("PageEditGuardians", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
