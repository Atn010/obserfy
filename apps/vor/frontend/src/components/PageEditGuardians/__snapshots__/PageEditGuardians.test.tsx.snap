// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageEditGuardians should render correctly 1`] = `
.emotion-17 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-17 {
    font-size: 20px;
  }
}

.emotion-16 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 700px;
  margin-left: auto;
  margin-right: auto;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: -8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-icon,rgba(0,0,0,0.6));
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 16px;
  height: 24px;
  width: 24px;
  margin-right: 4px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  text-rendering: optimizeLegibility;
  --webkitFontSmoothing: antialiased;
  --mozOsxFontSmoothing: grayscale;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  font-size: 16px;
  margin-bottom: 0;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    font-size: 14px;
  }
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 24px;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 300;
  line-height: heading;
  font-size: 2.5rem;
  line-height: 3rem;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  text-rendering: optimizeLegibility;
  --webkitFontSmoothing: antialiased;
  --mozOsxFontSmoothing: grayscale;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  margin-left: 16px;
  margin-right: 16px;
  margin-bottom: 32px;
  margin-top: 16px;
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  text-rendering: optimizeLegibility;
  --webkitFontSmoothing: antialiased;
  --mozOsxFontSmoothing: grayscale;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  font-size: 16px;
  margin-left: 16px;
  margin-right: 16px;
  margin-bottom: 8px;
}

@media screen and (min-width:40em) {
  .emotion-4 {
    font-size: 14px;
  }
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding: 8px;
  background-color: var(--theme-ui-colors-surface,#fff);
  border-radius: 8px;
  box-shadow: rgba(0,0,0,0.1) 0px 0px 1px,rgba(0,0,0,0.27) 0px 2px 4px -2px;
  border-radius: 0;
  padding: 0;
  overflow: hidden;
  margin-bottom: 8px;
  margin-left: 0;
  margin-right: 0;
}

@media screen and (min-width:40em) {
  .emotion-6 {
    border-radius: 8px;
  }
}

@media screen and (min-width:40em) {
  .emotion-6 {
    margin-left: 16px;
    margin-right: 16px;
  }
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  text-rendering: optimizeLegibility;
  --webkitFontSmoothing: antialiased;
  --mozOsxFontSmoothing: grayscale;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  font-size: 14px;
  margin: 16px;
}

.emotion-9 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  text-rendering: optimizeLegibility;
  --webkitFontSmoothing: antialiased;
  --mozOsxFontSmoothing: grayscale;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  font-size: 14px;
  line-height: 1;
}

.emotion-8 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0;
  height: 20px;
  width: 20px;
  margin-left: auto;
  margin-right: 8px;
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  text-rendering: optimizeLegibility;
  --webkitFontSmoothing: antialiased;
  --mozOsxFontSmoothing: grayscale;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  font-size: 16px;
  margin-left: 16px;
  margin-right: 16px;
  margin-top: 32px;
}

@media screen and (min-width:40em) {
  .emotion-11 {
    font-size: 14px;
  }
}

.emotion-15 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-bottom: 16px;
  padding-top: 8px;
}

.emotion-14 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  as: label;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-13 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-12 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 8px;
  color: inherit;
  background-color: transparent;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  width: auto;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  background-color: var(--theme-ui-colors-surface,#fff);
  font-size: 16px;
  padding: 12px;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  background-color: var(--theme-ui-colors-surface,#fff);
  width: 100%;
}

.emotion-12:hover,
.emotion-12:focus {
  border-color: var(--theme-ui-colors-primary,#00e399);
  outline: none;
}

.emotion-12::-webkit-input-placeholder {
  font-style: italic;
}

.emotion-12::-moz-placeholder {
  font-style: italic;
}

.emotion-12:-ms-input-placeholder {
  font-style: italic;
}

.emotion-12::placeholder {
  font-style: italic;
}

.emotion-12:disabled {
  opacity: 0.68;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  color: var(--theme-ui-colors-textDisabled,rgba(0,0,0,0.38));
}

<div>
  <div
    class="emotion-17"
  >
    <div
      class="emotion-16"
    >
      <a
        href="/dashboard/observe/students/profile?id=adhfakjldsf"
        style="display: block;"
      >
        <div
          class="emotion-2"
        >
          <div
            class="emotion-0"
          />
          <p
            class="emotion-1"
          >
            Student Profile
          </p>
        </div>
      </a>
      <h5
        class="emotion-3"
      />
      <p
        class="emotion-4"
      >
        Current guardians
      </p>
      <div
        class="emotion-6"
      >
        <p
          class="emotion-5"
        >
          No guardians set yet
        </p>
      </div>
      <div
        class="emotion-6"
      >
        <a
          data-cy="new-guardian"
          href="/dashboard/observe/students/profile/guardians/new?id=adhfakjldsf"
          style="display: block;"
        >
          <div
            class="emotion-9"
          >
            <p
              class="emotion-7"
            >
              Create new guardian
            </p>
            <div
              class="emotion-8"
            />
          </div>
        </a>
      </div>
      <p
        class="emotion-11"
      >
        Add existing guardian
      </p>
      <div
        class="emotion-15"
      >
        <label
          class="emotion-14"
        >
          <div
            class="emotion-13"
          >
            <input
              class="emotion-12"
              name="name"
              placeholder="Search"
              value=""
            />
          </div>
        </label>
      </div>
    </div>
  </div>
</div>
`;
